lib/posix-process/signal/deliver.c

```diff
@@ -0,0 +1,355 @@
+
+	uk_mutex_unlock(&pproc->signal->lock);
+
+	return 0;
+}
+
+/* Deliver pending signals of a given process. We deliver each
+ * pengind signal to the first thread that doesn't mask that
+ * signal.
+ */
+static int deliver_pending_proc(struct posix_process *proc,
+				struct uk_syscall_ctx *usc)
+{
+	struct posix_thread *thread, *threadn;
+	struct posix_thread *this_thread;
+	struct uk_signal *sig;
+	int handled_cnt = 0;
+	bool handled;
+
+	UK_ASSERT(proc);
+	UK_ASSERT(usc);
+
+	this_thread = tid2pthread(uk_gettid());
+
+	for (int i = 0; i < SIG_ARRAY_COUNT; i++) {
+		handled = false;
+
+		/* Skip if the process ignores this signal altogether */
+		if (IS_IGNORED(proc, i))
+			continue;
+
+		/* POSIX specifies that if a signal targets the
+		 * current process / thread, then at least one
+		 * signal for this process /thread must be
+		 * delivered before the syscall returns, as long as:
+		 *
+		 * 1. No other thread has that signal unblocked
+		 * 2. No other thread is in sigwait() for that signal (TODO)
+		 */
+		uk_process_foreach_pthread(proc, thread, threadn) {
+			if (thread->tid == this_thread->tid)
+				continue;
+
+			if (IS_MASKED(thread, i))
+				continue;
+
+			while ((sig = pprocess_signal_dequeue_p(proc, i))) {
+				do_deliver(thread, sig, usc);
+				uk_signal_free(proc->_a, sig);
+				handled = true;
+				handled_cnt++;
+			}
+			break;
+		}
+
+		/* Try to deliver to this thread */
+		if (!handled) {
+			if (IS_MASKED(this_thread, i))
+				continue;
+
+			while ((sig = pprocess_signal_dequeue_p(proc, i))) {
+				do_deliver(this_thread, sig, usc);
+				uk_signal_free(proc->_a, sig);
+				handled = true;
+				handled_cnt++;
+			}
+		}
+	}
+
+	return handled_cnt;
+}
+
+static int deliver_pending_thread(struct posix_thread *thread,
+				  struct uk_syscall_ctx *usc)
+{
+	struct posix_process *proc __maybe_unused;
+	struct uk_signal *sig;
+	int handled = 0;
+
+	proc = tid2pprocess(thread->tid);
+	UK_ASSERT(proc);
+
+	/* Deliver this thread's signals. SUS requires that RT signals
+	 * must be delivered starting from the lowest signal number.
+	 * Delivery order of standard signals is undefined. We deliver
+	 * all signals in order.
+	 */
+	for (int i = 1; i < SIG_ARRAY_COUNT; i++) {
+		if (!pprocess_signal_is_deliverable(thread, i))
+			continue;
+
+		while ((sig = pprocess_signal_dequeue_t(thread, i))) {
+			do_deliver(thread, sig, usc);
+			uk_signal_free(thread->_a, sig);
+			handled++;
+		}
+	}
+
+	return handled;
+}
+
+void uk_signal_deliver(struct uk_syscall_ctx *usc)
+{
+	struct posix_thread *pthread;
+	struct posix_process *pproc;
+	pid_t tid = uk_gettid();
+
+	UK_ASSERT(usc);
+
+	pthread = tid2pthread(tid);
+	UK_ASSERT(pthread);
+
+	pproc = pid2pprocess(uk_getpid());
+	UK_ASSERT(pproc);
+
+	/* If there's SIGKILL pending, kill the process right away */
+	if (IS_PENDING(pproc->signal->sigqueue, SIGKILL)) {
+		uk_pr_info("SIGKILL tid %d\n", tid);
+		uk_sigact_term(SIGKILL);
+		return;
```

Comment:
Should we reach this point? Maybe a `BUG_ON`?

Reply:
If the assumption is that we will only call this for the current process (which is) then yes this should not return.